### PR TITLE
[DF] Add support for systematic variations

### PR DESCRIPTION
**Pull request by eguiraud**
~~Test coverage is still not great, but feature-wise we should be there.~~

There might be edge cases that I have not thought of testing, but most things (including jitting) are implemented and have non-zero coverage.

### CODE REVIEW

**Review by Axel-Naumann — APPROVED** [review]
Need to continue with `tree/dataframe/inc/ROOT/RDF/RDefineBase.hxx` after the merge!

File: tree/dataframe/inc/ROOT/RDF/ColumnReaderUtils.hxx [hunk 1]
```diff
@@ -108,6 +131,15 @@ MakeColumnReaders(unsigned int slot, TTreeReader *r, TypeList<ColTypes...>, cons
    (void)r;
 }
 
+// dummy overload for for the case of no columns, to silence compiler warnings
+std::array<std::unique_ptr<RDFDetail::RColumnReaderBase>, 0> inline MakeColumnReaders(unsigned int, TTreeReader *,
+                                                                                      TypeList<>,
+                                                                                      const RColumnReadersInfo &,
+                                                                                      const std::string & = "nominal")
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/ColumnReaderUtils.hxx:138]: clang-format behaves better if `inline` comes first, fixed

File: tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx [hunk 2]
```diff
@@ -501,6 +509,55 @@ void JitDefineHelper(F &&f, const char **colsPtr, std::size_t colsSize, std::str
...
+                        std::string_view variationName, RLoopManager *lm,
+                        std::weak_ptr<RJittedVariation> *wkJittedVariation, RColumnRegister *colRegister,
+                        std::shared_ptr<RNodeBase> *prevNodeOnHeap) noexcept
+{
+   if (wkJittedVariation->expired()) {
+      // The branch of the computation graph that needed this jitted variation went out of scope between the type
+      // jitting was booked and the time jitting actually happened. Nothing to do other than cleaning up.
+      delete wkJittedVariation;
+      // colRegister must be deleted before prevNodeOnHeap because their dtor needs the RLoopManager to be alive
+      // and prevNodeOnHeap is what keeps it alive if the rest of the computation graph is already out of scope
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:524]: lifetime management is awkward because to make RDF's deferred jitting work we are putting stuff on the heap that normally wouldn't be, and as a consequence it could outlive everything else in the computation graph, which normally it could not. I'll see whether we can do better.

File: tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx [hunk 3]
```diff
@@ -501,6 +509,55 @@ void JitDefineHelper(F &&f, const char **colsPtr, std::size_t colsSize, std::str
...
+      delete colRegister;
+      delete prevNodeOnHeap;
+      return;
+   }
+
+   const ColumnNames_t inputColNames(colsPtr, colsPtr + colsSize);
+   delete[] colsPtr;
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:531]: the caller is long gone when this function is invoked: the arrays are allocated by compiled code, by `JitVariationHelper` is invoked in jitted code.

We used to pass `std::vector`s but that resulted in much longer jitting times than C-style arrays.

File: tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx [hunk 4]
```diff
@@ -501,6 +509,55 @@ void JitDefineHelper(F &&f, const char **colsPtr, std::size_t colsSize, std::str
...
+{
+   if (wkJittedVariation->expired()) {
+      // The branch of the computation graph that needed this jitted variation went out of scope between the type
+      // jitting was booked and the time jitting actually happened. Nothing to do other than cleaning up.
+      delete wkJittedVariation;
+      // colRegister must be deleted before prevNodeOnHeap because their dtor needs the RLoopManager to be alive
+      // and prevNodeOnHeap is what keeps it alive if the rest of the computation graph is already out of scope
+      delete colRegister;
+      delete prevNodeOnHeap;
+      return;
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:527]: YES! thanks! I'll put up another PR that fixes this in `Jit{Define,Filter,Variation}Helper` as soon as possible.

File: tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx [hunk 5]
```diff
@@ -501,6 +509,55 @@ void JitDefineHelper(F &&f, const char **colsPtr, std::size_t colsSize, std::str
...
+   }
+
+   const ColumnNames_t inputColNames(colsPtr, colsPtr + colsSize);
+   delete[] colsPtr;
+   std::vector<std::string> variedColNames(variedCols, variedCols + variedColsSize);
+   delete[] variedCols;
+   std::vector<std::string> tags(variationTags, variationTags + variationTagsSize);
+   delete[] variationTags;
+
+   auto jittedVariation = wkJittedVariation->lock();
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:537]: `wkJittedVariation` can either be "expired" (which we check at the beginning of the function) or contain a non-null pointer.

File: tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx [hunk 6]
```diff
@@ -501,6 +509,55 @@ void JitDefineHelper(F &&f, const char **colsPtr, std::size_t colsSize, std::str
...
 
+template <typename F>
+void JitVariationHelper(F &&f, const char **colsPtr, std::size_t colsSize, const char **variedCols,
+                        std::size_t variedColsSize, const char **variationTags, std::size_t variationTagsSize,
+                        std::string_view variationName, RLoopManager *lm,
+                        std::weak_ptr<RJittedVariation> *wkJittedVariation, RColumnRegister *colRegister,
+                        std::shared_ptr<RNodeBase> *prevNodeOnHeap) noexcept
+{
+   if (wkJittedVariation->expired()) {
+      // The branch of the computation graph that needed this jitted variation went out of scope between the type
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/InterfaceUtils.hxx:520]: absolutely, thanks! I'll fix it in a different PR because it needs fixing in `JitFilterHelper` and `JitDefineHelper` as well.

File: tree/dataframe/inc/ROOT/RDF/RAction.hxx [hunk 7]
```diff
@@ -64,8 +65,9 @@ class R__CLING_PTRCHECK(off) RAction : public RActionBase {
 
 public:
    RAction(Helper &&h, const ColumnNames_t &columns, std::shared_ptr<PrevNode> pd, const RColumnRegister &colRegister)
-      : RActionBase(pd->GetLoopManagerUnchecked(), columns, colRegister), fHelper(std::forward<Helper>(h)),
-        fPrevNodePtr(std::move(pd)), fPrevNode(*fPrevNodePtr), fValues(GetNSlots()), fIsDefine()
+      : RActionBase(pd->GetLoopManagerUnchecked(), columns, colRegister, pd->GetVariations()),
+        fHelper(std::forward<Helper>(h)), fPrevNodePtr(std::move(pd)), fPrevNode(*fPrevNodePtr), fValues(GetNSlots()),
+        fIsDefine()
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RAction.hxx:70]: we don't need `fIsDefine()` at all, we are happy with the indeterminate default initialization as we set the values in the ctor body. I removed it.

File: tree/dataframe/inc/ROOT/RDF/RActionBase.hxx [hunk 8]
```diff
@@ -84,6 +87,10 @@ public:
    virtual std::unique_ptr<RMergeableValueBase> GetMergeableValue() const = 0;
 
    virtual ROOT::RDF::SampleCallback_t GetSampleCallback() = 0;
+
+   const std::vector<std::string> &GetVariations() const { return fVariations; }
+
+   virtual std::unique_ptr<RActionBase> MakeVariedAction(std::vector<void *> &&results) = 0;
```
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/inc/ROOT/RDF/RActionBase.hxx:93]: Users never have to inherit from `RActionBase`. Implementing custom actions is done by inheriting from `RActionImpl` (with CRTP).
